codecs/av1_packet.go

```diff
@@ -0,0 +1,138 @@
+
+// Payload fragments a AV1 packet across one or more byte arrays
+func (p *AV1Payloader) Payload(mtu uint16, payload []byte) [][]byte {
+	return nil
+}
+
+// AV1Packet represents a depacketized AV1 RTP Packet
+//
+//  0 1 2 3 4 5 6 7
+// +-+-+-+-+-+-+-+-+
+// |Z|Y| W |N|-|-|-|
+// +-+-+-+-+-+-+-+-+
+//
+// https://aomediacodec.github.io/av1-rtp-spec/#44-av1-aggregation-header
+type AV1Packet struct {
+	// Z: MUST be set to 1 if the first OBU element is an
+	//    OBU fragment that is a continuation of an OBU fragment
+	//    from the previous packet, and MUST be set to 0 otherwise.
+	Z byte
+
+	// Y: MUST be set to 1 if the last OBU element is an OBU fragment
+	//    that will continue in the next packet, and MUST be set to 0 otherwise.
+	Y byte
+
+	// W: two bit field that describes the number of OBU elements in the packet.
+	//    This field MUST be set equal to 0 or equal to the number of OBU elements
+	//    contained in the packet. If set to 0, each OBU element MUST be preceded by
+	//    a length field. If not set to 0 (i.e., W = 1, 2 or 3) the last OBU element
+	//    MUST NOT be preceded by a length field. Instead, the length of the last OBU
+	//    element contained in the packet can be calculated as follows:
+	// Length of the last OBU element =
+	//    length of the RTP payload
+	//  - length of aggregation header
+	//  - length of previous OBU elements including length fields
+	W byte
+
+	// N: MUST be set to 1 if the packet is the first packet of a coded video sequence, and MUST be set to 0 otherwise.
+	N byte
+
+	OBUs [][]byte
+
+	// Buffer for fragmented OBU. If Unmarshal is called on a RTP Packet
+	// that doesn't contain a fully formed OBU
+	obuBuffer []byte
+}
+
+func (p *AV1Packet) pushOBUElement(isFirstOBUFragment *bool, obuElement []byte) {
+	if *isFirstOBUFragment {
+		*isFirstOBUFragment = false
+		// Discard pushed because we don't have a fragment to combine it with
+		if p.obuBuffer == nil {
+			return
+		}
+		obuElement = append(p.obuBuffer, obuElement...)
+		p.obuBuffer = nil
+	}
+	p.OBUs = append(p.OBUs, obuElement)
+}
+
+// Unmarshal parses the passed byte slice and stores the result in the VP8Packet this method is called upon
+func (p *AV1Packet) Unmarshal(payload []byte) ([]byte, error) {
+	if payload == nil {
+		return nil, errNilPacket
+	} else if len(payload) < 2 {
+		return nil, errShortPacket
+	}
+
+	p.Z = (payload[0] & zMask) >> zBitshift
+	isFirstOBUFragment := p.Z != 0
+	p.Y = (payload[0] & yMask) >> yBitshift
+	p.N = (payload[0] & nMask) >> nBitshift
+	p.W = (payload[0] & wMask) >> wBitshift
+
+	if isFirstOBUFragment && p.N != 0 {
+		return nil, errIsKeyframeAndFragment
+	}
+
+	currentIndex := uint(1)
+	p.OBUs = [][]byte{}
+
+	var (
+		obuElementLength, bytesRead uint
+		err                         error
+	)
+	for i := 1; ; i++ {
+		if currentIndex == uint(len(payload)) {
+			break
+		}
+
+		// If W bit is set the last OBU Element will have no length header
+		if byte(i) == p.W {
```

Comment:
I remember W can't be greater than 3 and should fail.

Reply:
Added `errWGreaterThenThree`